### DISCUSSION TITLE
Are query params supposed to persist after `router.push`

### DISCUSSION THREAD

**Discussion by tylersayshi** [47ng/nuqs#643] [Q&A] [answered]
Let's say I am on `/about?name=foo` then I run `router.push('/elsewhere')` from next's router.

Currently, I am seeing this navigate me to `/elsewhere?name=foo`... Is this expected? My understanding was that query parameters would only persist if explicitly passed to `router.push('/elsewhere' + queryParams)`. 

Am I misunderstanding how this should work?

If my understanding is correct, what might cause nuqs to persist query params across paths?

If I am incorrect, could you recommend a way to clear the query parameters on this `router.push` event.

Thanks in advance! nuqs is great :)

**Comment by tylersayshi** [comment]
So I am observing two steps here

1. I call `router.push('/elsewhere')`. This correctly triggers the navigate event to just /elsewhere
2. a `router.replace` is triggered by this line here: https://github.com/47ng/nuqs/blob/v1.19.1/packages/nuqs/src/sync.ts#L120

Is this expected behavior and can it be disabled?

**Reply by tylersayshi** [reply]
as I console.log on `navigation.addEventListener('navigate', ...)` I see:

1. push to /elsewhere
2. replace from nuqs with query params that were used on /about
3. a replace that removes one of the 4 query params that got copied from the last page (There's no apparent pattern to why this one was removed)
4. The one that was removed in step 3 re-appears & the first calls to setState (from useQueryState) on the new page add two new query param items

This _seems_ like a bug afaict

**Reply by franky47** [reply, marked as answer]
That's wild indeed. If you're able, could you give v2 a try? The whole sync thing has been removed there (it was only necessary for some unstable versions of Next.js' shallow routing).

```
pnpm add https://pkg.pr.new/nuqs@459
```

Docs for migration (only hard requirement if you're not using deprecated things is to be on `next@>=14.1.2`):
https://github.com/47ng/nuqs/pull/459

**Reply by tylersayshi** [reply]
With initial testing - this just works :) Thanks so much for the quick response. 

Is there anything I should look out for with using v2 now? Or would you consider it in a fairly stable state that is just waiting on react 19 to be released?

And I didn't notice the Provider requirement you tweeted about yet in the migration notes, I assume that just hasn't been added yet then?

**Reply by franky47** [reply]
Yeah the adapters PR is a separate one (#644), based on top of v2, but if it's stable enough before Next 15 / React 19 comes out, it'll get merged into v2 and so the main breaking change will be adding the adapters.

**Reply by tylersayshi** [reply]
Do you think you could publish a v2 alpha or beta version to npm?

**Reply by tylersayshi** [reply]
Ah, I am actually able to pin to a commit with `https://pkg.pr.new/nuqs@97a1ea8`

This is good enough for what I'd need, but it might be good to publish v2 to npm before stable just for others to try :) Thanks again!

**Reply by franky47** [reply]
Yes, there should be beta versions for v2 once it gets merged before landing on `latest`, semantic-release is a bit of a pain to use in this context though, pkg.pr.new works better for WIP stuff. Maybe I'll enable the `rc` dist-tag to follow Next & React and trigger manual updates on it.

**Reply by firatciftci** [reply]
Going against the bug report here, but, is there any way to actually _persist_ the query parameters after a `router.push()` or a `<Link href="" />`? I would love to be able to set an option in `useQueryState/s()` (e.g., `persist`; default is `false`, and `true` would enable cross-page navigation persistance) so that navigating to another page does not flush the search query state.

For example, I have a sidebar with multiple items, all being `<Link />`s, and above them is an `<input />` that sets the `search` query parameter on submit. The search query parameter works as a filter on the `name` property of the sidebar items, so is pretty integral to the application's main logic. However, when I click on an item and thus navigate to `/item/[id]`, the search query parameter gets reset to `null`.

To remedy this, I need to add a less-than-ideal workaround, as such:
```tsx
import Link from "next/link";
import { useQueryState, parseAsString } from "nuqs":

export default function Item({ name, id }: { name: string, id: string }) {
  const [search] = useQueryState("search", parseAsString.withOptions({ history: "push" });

  return <Link href={`/item/${id}${search !== null ? `?search=${search}` : ""}>{name}</Link>
}
```

Of course, not the biggest problem to have, but you lose out on all type-safety, compounded with the fact that the problem gets worse as you have more search query parameters to keep track of across page changes.

**Reply by franky47** [reply]
@firatciftci how would that work in relation to subsequent navigations?

Eg: 
1. Start at `/`, set `/?search=foo` by typing text in the search input
2. Click item, navigate to `/item/foobar?search=foo` (persisted search param)
3. Navigate to another unrelated page (eg: `/terms-of-service`), would you keep `?search=foo`? Probably not, if there is nothing there that requires it.

I agree that there is a missing piece for type-safe navigation.

TanStack Router solved this problem by having a centralised route declaration, the approach with nuqs was to scope search params to components that can be moved around pages, avoiding tying search params to a particular page, but it makes safe routing difficult by having to know which page actually uses which nuqs-enabled components ahead of time.

Some other projects have attempted to replicate the TanStack way in Next, see [`next-safe-navigation`](https://github.com/lukemorales/next-safe-navigation).

**Reply by firatciftci** [reply]
That seems like a super cool package, and I will definitely check it out; thanks! In terms of the matter at hand though, items 1 and 2 in your example list I agree with, but have issues with item 3: you are correct that it would be odd to expect search parameters to persist _always_, but I think it is also equally valuable to support applications where an _always_ is required; an application should not need/have to bail out on `nuqs` for a specific case.

For the proposed `persist` flag, I would expect that only those who actually need it to show up in _all_ pages would set it to `true`. For the majority of expected use cases, the default `false` should cover the basis, which I think is valid and perfectly enough for a flag-based feature.

If you would like a more holistic solution to this (without implementing a full or TanStack-like router [which, to be perfectly honest with you, I would love to see _you_ tackle, as `nuqs` has become my favorite Next.js/React package]), the following could/would work:
```tsx
import { useQueryState, parseAsString } from "nuqs";

const [search] = useQueryState("search", parseAsString.withOptions({ history: "push", persist: ["/item/[slug]"] }));
```

In this case, the value for the `persist` key could be `boolean` or `Array<string>`; the former would allow search parameter persistance on _all_ cross-navigation cases, and the latter would basically act as a whitelist for this purpose. (I know the _most ideal_ for this scenario would be to have type-safety in the string array too, but, very good is sometimes alright compared to perfect too. 🙂)

**Reply by firatciftci** [reply]
And, perhaps the `experimental.typedRoutes` feature of Next.js ([see here](https://nextjs.org/docs/app/building-your-application/configuring/typescript#statically-typed-links)) could provide a basis for type-safety for this `Array<string>` value; I would not dare ask for a crazy complex feature, as the most basic implementation (the first round) is perfectly acceptable, but I think there is something to look at here!

P.S.: If you need a second pair of hands, I would love to see how I could do this; let me know!

**Reply by franky47** [reply]
Point 3. and having search params persist across all the app have something in common though: there needs to be, somewhere, something that does this kind of persistence.

If you were to put this flag as an option in useQueryState(s), this would only work on pages where they are mounted. In v2, there'll be adapters wrapping the entire application, so centralising what params to persist might make more sense to be defined at the adapter level (and we're getting close to a global route definition à la TSR).

I'll have to mull over this to find a good way to go further than isolated hooks. It's unlikely that anything will come off it in v2 though (already a lot to do and Next.js conf is closing in fast).
